core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +92,43 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    /* add waiter address to atomic wait list */
+    AtomicWaitAddressNode *wait_address_node;
+
+    if (!(wait_address_node =
+              wasm_runtime_malloc(sizeof(AtomicWaitAddressNode)))) {
+        return;
+    }
+    wait_address_node->addr = key;
+    bh_assert(bh_list_insert(user_data, wait_address_node) == BH_LIST_SUCCESS);
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    bh_list atomic_wait_address_list_head;
+    bh_list *const atomic_wait_address_list = &atomic_wait_address_list_head;
+    static korp_mutex atomic_wait_address_list_lock;
+    AtomicWaitAddressNode *node;
+
+    bh_list_status ret = bh_list_init(atomic_wait_address_list);
+    bh_assert(ret == BH_LIST_SUCCESS);
+
+    os_mutex_lock(&atomic_wait_address_list_lock);
+    bh_hash_map_traverse(wait_map, create_list_of_waiter_addresses,
```

Comment:
for this traversal, we should probably lock a `shared_memory_list_lock` otherwise the traversal might be invalidated half way through?